include/osmium/builder/osm_object_builder.hpp

```diff
@@ -362,32 +372,53 @@ namespace osmium {
-                assert(!m_comment && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
-                m_comment = reserve_space_for<osmium::ChangesetComment>();
-                new (m_comment) osmium::ChangesetComment{date, uid};
+                assert(m_comment_offset == SIZE_MAX && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
+                
+                // Store offset instead of pointer to handle buffer reallocation
+                m_comment_offset = buffer().written() - buffer().committed();
+                
+                auto* comment = reserve_space_for<osmium::ChangesetComment>();
+                new (comment) osmium::ChangesetComment{date, uid};
                 add_size(sizeof(ChangesetComment));
-                add_user(*m_comment, user, std::strlen(user));
+                
+                auto* comment_ptr = get_comment_ptr();
+                add_user(*comment_ptr, user, std::strlen(user));
             }
 
             void add_comment_text(const char* text) {
-                assert(m_comment && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
-                osmium::ChangesetComment& comment = *m_comment;
-                m_comment = nullptr;
+                assert(m_comment_offset != SIZE_MAX && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
+                
+                // Get fresh pointer each time to handle buffer reallocation
+                auto* comment_ptr = get_comment_ptr();
+                osmium::ChangesetComment& comment = *comment_ptr;
+                
+                // Invalidate offset to ensure right adding order
+                m_comment_offset = SIZE_MAX;
+                
                 add_text(comment, text, std::strlen(text));
             }
 
             void add_comment_text(const std::string& text) {
-                assert(m_comment && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
-                osmium::ChangesetComment& comment = *m_comment;
-                m_comment = nullptr;
+                assert(m_comment_offset != SIZE_MAX && "You have to always call both add_comment() and then add_comment_text() in that order for each comment!");
+                
+                // Get fresh pointer each time to handle buffer reallocation
+                auto* comment_ptr = get_comment_ptr();
+                osmium::ChangesetComment& comment = *comment_ptr;
+                
+                // Invalidate offset to ensure right adding order
+                m_comment_offset = SIZE_MAX;
+                
                 add_text(comment, text.c_str(), text.size());
             }
 
+            bool has_open_comment() const noexcept {
```

Comment:
You wrote this function but are not using it. I think it is a good idea, so use it. But I think it should be `private`.